### ISSUE TITLE
Pattern matching spans of chars against constant strings

### ISSUE THREAD

**Issue by brianrourkeboll** [fsharp/fslang-suggestions#1351]
## Proposal

I propose that we allow pattern-matching values of type `ReadOnlySpan<char>` and `Span<char>` against constant strings.

```fsharp
match "abc123".AsSpan 1 with
| "bc123"  -> printfn "Matches."
| "abc123" -> printfn "Doesn't match."
| "a"      -> printfn "Doesn't match."
| _        -> printfn "Doesn't match."
```

```fsharp
let [<Literal>] SomeConstant = "abc123"

let f (span : ReadOnlySpan<char>) =
    match span with
    | SomeConstant -> printfn "Matches."
    | _            -> printfn "Doesn't match."
```

```fsharp
let span = "abc123xyz".AsSpan ()
let lastDigit = span.LastIndexOfAnyInRange ('0', '9')

match span.Slice lastDigit with
| ""     -> printfn "Ends with a digit."
| suffix -> printfn $"Ends with '{suffix.ToString ()}'."
```

Compare the [equivalent C# proposal](https://github.com/dotnet/csharplang/blob/6ac33c5f93c3fdd4d1addd68f2e8f99a79d0c191/proposals/csharp-11.0/pattern-match-span-of-char-on-string.md) (implemented in C# 11).

We would do this by first automatically generating calls to `System.MemoryExtensions.AsSpan` on a constant string being matched against, and then by passing the resulting span and the input span to the appropriate `System.MemoryExtensions.SequenceEqual` overload.

## The existing way of approaching this problem in F# is...

- Use string slicing instead of span slicing. This allocates a new string.

  ```fsharp
  match "abc123"[1..] with // Or "abc123".Substring 1, etc.
  | "bc123"  -> printfn "Matches."
  | "abc123" -> printfn "Doesn't match."
  | "a"      -> printfn "Doesn't match."
  | _        -> printfn "Doesn't match."
  ```

- Use a series of `if`-`then`-`else` expressions instead of pattern-matching.

  ```fsharp
  let slice = "abc123".AsSpan 1

  if slice.SequenceEqual "bc123" then
      printfn "Matches."
  elif slice.SequenceEqual "abc123" then
      printfn "Doesn't match."
  elif slice.SequenceEqual "a" then
      printfn "Doesn't match."
  else
      printfn "Doesn't match."
  ```

## Notes

- We would not change the type inference rules for string patterns — the inferred type for the pattern input of a string literal pattern would still be `string`.

- As in the C# implementation, I think we would want a compile-time error for matching against `null`:

  ```fsharp
  let [<Literal>] NullLiteral : string = null
  
  match "abc123".AsSpan (2, 3) with
  | (null : string) (* This should be an error at build-time. *) -> …
  | NullLiteral (* This should be an error at build-time. *) -> …
  | _ -> …
  ```

- Any named pattern bindings should still have the input (span) type:

  ```fsharp
  match "asdf".AsSpan () with
  | "asdf" &  span -> printfn "`span` is a `ReadOnlySpan<char>` here."
  | "asdf" as span -> printfn "`span` is a `ReadOnlySpan<char>` here."
  | span           -> printfn "`span` is a `ReadOnlySpan<char>` here."
  ```

## Pros and cons

**The advantages of making this adjustment to F# are**

- It brings the ergonomics of working with spans closer to the ergonomics already enjoyed by strings. This would help improve the legibility of performance-sensitive code that uses spans to avoid unnecessary string copying and allocations.

**The disadvantages of making this adjustment to F# are**

- None, except that we're moving the complexity from the programmer to the language.

## Alternatives

- Type-directed `[…]` syntax (#1086) could theoretically subsume this, although when the underlying thing being sliced is a string, I think allowing matching against string literals still has a readability advantage. C# supports both, for what it's worth:

  ```csharp
  _ = "abc123".AsSpan(1, 3) switch
  {
      "bc1" => "Matches.",
      ['b', 'c', '1'] => "Also matches.",
      _ => "…"
  };
  ```

## Extra information

**Estimated cost (XS, S, M, L, XL, XXL):**

- S — to the best of my understanding, since such code does not currently compile, and since I don't know of anything that making this addition to the compiler would later preclude.

**Related suggestions:** (put links to related suggestions here)

- This could potentially be viewed as an extension of #849 (see discussion of `op_Implicit` in [FS-1093](https://github.com/fsharp/fslang-design/blob/2e9d73667ef576cd0c3dad7b4bb5c9b06e578b3b/FSharp-6.0/FS-1093-additional-conversions.md)).

## Affidavit (please submit!)

Please tick these items by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [StackOverflow](http://stackoverflow.com)) and I have searched StackOverflow for discussions of this issue
* [x] This is a language change and not purely a tooling change (e.g. compiler bug, editor support, warning/error messages, new warning, non-breaking optimisation) belonging to [the compiler and tooling repository](https://github.com/dotnet/fsharp)
* [x] This is not something which has obviously "already been decided" in previous versions of F#. If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by vzarytovskii** [comment]
I am personally not a huge fan of introducing intrinsics to the compiler it should know of to be able to optimize it. Implicit conversion _looks_ better tbh, but will involve additional hidden conversions/allocations. Also, confusing from the usability/type system perspective - "Why do I compare Span to string"?

**Comment by brianrourkeboll** [comment]
> will involve additional hidden conversions/allocations

Hmm, it would indeed involve hidden invocation of `op_Implicit` and creation of `ReadOnlySpan<char>`s, but it wouldn't involve any heap allocations.

> Also, confusing from the usability/type system perspective - "Why do I compare Span to string"?

While I can sympathize with this perspective somewhat, once you know what spans _are_ and are _for_, then it's annoying _not_ to be able to mix and match spans created from strings and strings themselves — C# enables this on purpose by using `op_Implicit` to autoconvert strings to spans in as many scenarios as possible. I of course wouldn't argue for more implicit conversions in general, but I think the benefit would outweigh the potential confusion in this specific scenario.

Separately, while it's not exactly the same, there is already some precedent in the language for syntactic crossover between string-like things and array-like things:

```fsharp
// Looks like a string...
match "abc"B with
// ...But it's actually an array.
| [|0x61uy; 0x62uy; 0x63uy|] -> printfn "Matches."
| _ -> printfn "Doesn't match."
```

**Comment by baronfel** [comment]
From a design perspective, it's also nice as a user when the _easy_ thing is also the fast/correct thing.

**Comment by vzarytovskii** [comment]
> > will involve additional hidden conversions/allocations
> 
> 
> 
> Hmm, it would indeed involve hidden invocation of `op_Implicit` and creation of `ReadOnlySpan<char>`s, but it wouldn't involve any heap allocations.
> 
> 

Yeah, I meant if we do type directed the other way (span -> string) for some cases, where we can't (?) do AsSpan.


> 
> > Also, confusing from the usability/type system perspective - "Why do I compare Span to string"?
> 
> 
> 
> While I can sympathize with this perspective somewhat, once you know what spans _are_ and are _for_, then it's annoying _not_ to be able to mix and match spans created from strings and strings themselves — C# enables this on purpose by using `op_Implicit` to autoconvert strings to spans in as many scenarios as possible. I of course wouldn't argue for more implicit conversions in general, but I think the benefit would outweigh the potential confusion in this specific scenario.
> 
> 

Yeah, it can just be somewhat confusing for some people, I suppose.

**Comment by vzarytovskii** [comment]
> From a design perspective, it's also nice as a user when the _easy_ thing is also the fast/correct thing.

I suppose it's a tradeoff - do exactly what user wrote (or close to it) versus try and rewrite/optimize things. We already do both, so no really huge directional change.

I suppose it's up to @dsyme now - whether we want to do more with spans (and byreflikes in general), since it's clearly where Roslyn is going (and clr in general).

**Comment by charlesroddie** [comment]
I don't understand this semantically. `"bc123"` appears to be a `string`, and so the match appears to be equality, but a `string` is not a `ReadOnlySpan` so equality doesn't make sense.

**Comment by brianrourkeboll** [comment]
> `"bc123"` appears to be a `string`, and so the match appears to be checking for equality, but a `string` is not a `ReadOnlySpan` so equality doesn't make sense.

Yeah, I get that it "is" not a string, and that it might not "feel" F#-y if you focus on the current _type_ implications of string literals rather than viewing them as a _shape_ or _pattern_ of `char`s to match against — see my response above (https://github.com/fsharp/fslang-suggestions/issues/1351#issuecomment-1960064255):

> While I can sympathize with this perspective somewhat, once you know what spans _are_ and are _for_, then it's annoying _not_ to be able to mix and match spans created from strings and strings themselves — C# enables this on purpose by using `op_Implicit` to autoconvert strings to spans in as many scenarios as possible. I of course wouldn't argue for more implicit conversions in general, but I think the benefit would outweigh the potential confusion in this specific scenario.

The whole point of spans is to serve as windows into contiguous stretches of memory — and a string literal is just special syntax for a stretch of contiguous `char`s.

Sure, when a string literal is used as an object, an object is created to point at that stretch of `char`s, but — as the treatment of strings and spans in C# string slicing and pattern-matching as well as [the design of `System.String` itself in the BCL](https://github.com/dotnet/runtime/blob/79dd9bae9bb881eb716b608577c4cedc6c9cba72/src/libraries/System.Private.CoreLib/src/System/String.cs#L359-L362) show — strings and spans _are_ meant to be treated as more or less interchangeable, including in pattern-matching.

I do understand that the usage of `op_Implicit`, etc., would probably not be how it would have been done if the idea of spans had existed in .NET (and thus F#) from day one, and that it doesn't fit perfectly in the F# tradition, but it seems clear that testing equality between a span of `char`s and a string _is_ meant to be a meaningful operation from the .NET perspective.

So in my mind the change here doesn't imply some new _type_ equivalence between spans and strings, but rather specifically enables treating string literals as _patterns_ representing spans of characters against which, well `Span`s or `ReadOnlySpan`s of characters can be matched.

**Comment by T-Gro** [comment]
The motivation here is well explained.

It is true that ROS<char> is getting on par with string in BCL via overloads, and it would be good to cover ` "somestring" is ROS<char> ` via a higher-level design covering more areas of the language (method arguments, matching, interpolation, ..). A design that does not leak to programmers and internally chooses to pick `ROS<char>` ; and yet does not feel ad-hocy.

A symptom of a missing higher-level concept is that even the motivating samples had to use `AsSpan()`, which I consider a leaking implementation detail (a detail flowing in from decisions done in .NET).

I will mark this suggestion as approved in principle, but the design might be better of under an unified `ReadOnlySpan<char>` umbrella.
